protoc-gen-gogo/generator/generator.go

```diff
@@ -2343,12 +2343,19 @@ func (g *Generator) generateMessage(message *Descriptor) {
 
 	g.P("func (m *", ccTypeName, ") XXX_Marshal(b []byte, deterministic bool) ([]byte, error) {")
 	g.In()
-	g.P("if m, ok := (interface{})(m).(proto.Marshaler); ok {")
-	g.In()
-	g.P("return m.Marshal()")
-	g.Out()
-	g.P("}")
-	g.P("return xxx_messageInfo_", ccTypeName, ".Marshal(b, m, deterministic)")
+	if gogoproto.IsMarshaler(g.file.FileDescriptorProto, message.DescriptorProto) ||
```

Comment:
If we do a compile time check here, then we should do the same for the XXX_Unmarshal method.  We should just try to be consistent.